src/interp.rs

```diff
@@ -19,153 +19,483 @@
+                        zi = 0;
+                    }
+                }
+            }
+
+            self.zi = (self.zi + frames) % self.delay;
         }
     }
+}
 
-    pub fn get_factor(&self) -> usize {
-        self.factor
+/// A trait to be generic over number of channels in a of frame
+trait Frame: Sized + Copy {
+    const CHANNELS: usize;
+
+    fn scale_add(&mut self, other: &Self, coeff: f32);
+    fn from_planar(slice: &[f32], stride: usize) -> Self;
+}
+
+type MonoFrame32 = [f32; 1];
+type StereoFrame32 = [f32; 2];
+type SurroundFrame32 = [f32; 6];
+
+impl Frame for MonoFrame32 {
+    const CHANNELS: usize = 1;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
     }
 
-    pub fn reset(&mut self) {
-        // TODO: Use slice::fill() once stabilized
-        for v in &mut *self.z {
-            *v = 0.0;
-        }
-        self.zi = 0;
+    #[inline(always)]
+    fn from_planar(slice: &[f32], _stride: usize) -> Self {
+        [slice[0]]
     }
+}
 
-    pub fn process(&mut self, src: &[f32], dst: &mut [f32]) {
-        assert!(src.len().checked_mul(self.factor) == Some(dst.len()));
-        assert!(self.z.len() == self.delay * self.channels as usize);
-        assert!(self.filter.len() == self.factor);
-        assert!(self.zi < self.delay);
-
-        if src.is_empty() {
-            return;
-        }
-
-        let frames = src.len() / self.channels as usize;
-
-        for (src, (dst, z)) in src.chunks_exact(frames).zip(
-            dst.chunks_exact_mut(frames * self.factor)
-                .zip(self.z.chunks_exact_mut(self.delay)),
-        ) {
-            let mut zi = self.zi;
-
-            for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
-                // Add sample to delay buffer
-                //
-                // TODO Ringbuffer without bounds checks for z/zi
-                //
-                // Safety: zi is checked to be between 0 and self.delay
-                *unsafe { z.get_unchecked_mut(zi) } = *src;
-
-                // Apply coefficients
-                for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
-                    let mut acc = 0.0;
-                    for (c, index) in &*filter.coeff {
-                        let mut i = zi as i32 - *index as i32;
-                        if i < 0 {
-                            i += self.delay as i32;
-                        }
-                        // Safety: zi is checked to be between 0 and self.delay
-                        acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
+impl Frame for StereoFrame32 {
+    const CHANNELS: usize = 2;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
+        self[1] += other[1] * coeff;
+    }
+
+    #[inline(always)]
+    fn from_planar(slice: &[f32], stride: usize) -> Self {
+        [slice[0], slice[stride]]
```

Comment:
Is the bounds check for `slice[stride]` optimized away?

Reply:
Unfortunately not. Any ideas on how to achieve avoiding unsafe?

Reply:
Not directly, for now add some unsafe code for that. I assume it makes a measurable difference?

Reply:
I'll check.

Reply:
The difference is within the margin of error, so I suggest let's leave it with safety-checked for now.